### ISSUE TITLE
Radix code must not use pmd_huge() etc. on kernel page tables when HUGETLBFS=n

### ISSUE THREAD

**Issue by shenki** [linuxppc/issues#237]
The petitboot kernel (v5.0, powernv, with CONFIG_STRICT_KERNEL_RWX=y, gcc 7.3, binutils 2.30) was crashing with the following. @mpe helped with some initial investigation.

```
[   37.048138] Run /init as init process
[   37.117882] udevd[3201]: starting version 3.2.5
[   37.125604] udevd[3202]: starting eudev-3.2.5
[   37.354371] tg3 0004:01:00.0 enP4p1s0f0: renamed from eth0
[   37.355576] kernel tried to execute exec-protected page (c000000002bff4f0) -exploit attempt? (uid: 0)
[   37.356610] BUG: Unable to handle kernel instruction fetch
[   37.357724] Faulting instruction address: 0xc000000002bff4f0
cpu 0x2f: Vector: 400 (Instruction Access) at [c000001ff43ab6a0]
    pc: c000000002bff4f0
    lr: c000000000c001c0: kobject_rename+0x2c/0x160
    sp: c000001ff43ab930
   msr: 9000000010009033
  current = 0xc000001ff1049180
  paca    = 0xc000001ffffdb800	 irqmask: 0x03	 irq_happened: 0x01
    pid   = 3222, comm = udevd
Linux version 5.0.0 (joel@ka5) (gcc version 7.3.0 (Ubuntu 7.3.0-27ubuntu1~18.04)) #22 SMP Thu Apr 11 00:07:31 AEST 2019
enter ? for help
[link register   ] c000000000c001c0 kobject_rename+0x2c/0x160
[c000001ff43ab930] 5deadbeef0000200 (unreliable)
[c000001ff43ab9e0] c0000000005055b4 device_rename+0x74/0xd8
[c000001ff43aba30] c000000000aff910 dev_change_name+0x1c8/0x354
[c000001ff43abb10] c000000000b262bc dev_ifsioc+0x314/0x358
[c000001ff43abb70] c000000000b26950 dev_ioctl+0x4cc/0x4f0
[c000001ff43abbc0] c000000000ad9a88 sock_do_ioctl.constprop.14+0x12c/0x16c
[c000001ff43abc30] c000000000ada0a4 sock_ioctl+0x100/0x37c
[c000001ff43abcf0] c00000000016c3d0 vfs_ioctl+0x3c/0x60
[c000001ff43abd10] c00000000016ca5c do_vfs_ioctl+0xc0/0x7a0
[c000001ff43abdb0] c00000000016d184 ksys_ioctl+0x48/0x74
[c000001ff43abe00] c00000000016d1d0 sys_ioctl+0x20/0x78
[c000001ff43abe20] c00000000000b39c system_call+0x5c/0x70
--- Exception: c01 (System Call) at 00007fffb96121a4
SP (7fffd19f1230) is in userspace
```

```
di %lr 20
c000000000c001c0  7c7f1b79	mr.     r31,r3
c000000000c001c4  4182005c	beq     c000000000c00220	# kobject_rename+0x8c/0x160
c000000000c001c8  e93f0018	ld      r9,24(r31)
c000000000c001cc  2da90000	cmpdi   cr3,r9,0
c000000000c001d0  419e0050	beq     cr7,c000000000c00220	# kobject_rename+0x8c/0x160
c000000000c001d4  3c800060	lis     r4,96
c000000000c001d8  608400c0	ori     r4,r4,192
c000000000c001dc  49fff051	bl      c000000002bff22c
c000000000c001e0  7c7b1b79	mr.     r27,r3
c000000000c001e4  4082005c	bne     c000000000c00240	# kobject_rename+0xac/0x160
c000000000c001e8  3b800000	li      r28,0
c000000000c001ec  39a00000	li      r13,0
c000000000c001f0  3bc0fff4	li      r30,-12
c000000000c001f4  7da3eb78	or      r3,r13,r29
c000000000c001f8  4b524b71	bl      c000000000124d68	# kfree_const+0x8/0x4c
c000000000c001fc  60000000	nop
c000000000c00200  7f83e378	mr      r3,r28
c000000000c00204  4b552201	bl      c000000000152404	# kfree+0x8/0x1ec
c000000000c00208  60000000	nop
c000000000c0020c  7f63db78	mr      r3,r27
c000000000c00210  4b5521f5	bl      c000000000152404	# kfree+0x8/0x1ec
c000000000c00214  60000000	nop
c000000000c00218  7fe3fb78	mr      r3,r31
c000000000c0021c  4bfff4a1	bl      c000000000bff6bc	# kobject_put+0x8/0xfc
c000000000c00220  e9210078	ld      r9,120(r1)
c000000000c00224  e94d0b68	ld      r10,2920(r13)
c000000000c00228  7d295279	xor.    r9,r9,r10
c000000000c0022c  39400000	li      r10,0
c000000000c00230  7fc3f378	mr      r3,r30
c000000000c00234  418200b8	beq     c000000000c002ec	# kobject_rename+0x158/0x160
c000000000c00238  4b47b8a1	bl      c00000000007bad8	# __stack_chk_fail+0x8/0x24
c000000000c0023c  60000000	nop
```

It's the call to `kstrup_const()`

```
2f:mon> di c000000000c001b0 10
c000000000c001b0  e92d0b68	ld      r9,2920(r13)
c000000000c001b4  f9210078	std     r9,120(r1)
c000000000c001b8  39200000	li      r9,0
c000000000c001bc  49fff335	bl      c000000002bff4f0
c000000000c001c0  7c7f1b79	mr.     r31,r3
c000000000c001c4  4182005c	beq     c000000000c00220	# kobject_rename+0x8c/0x160
c000000000c001c8  e93f0018	ld      r9,24(r31)
c000000000c001cc  2da90000	cmpdi   cr3,r9,0
c000000000c001d0  419e0050	beq     cr7,c000000000c00220	# kobject_rename+0x8c/0x160
c000000000c001d4  3c800060	lis     r4,96
c000000000c001d8  608400c0	ori     r4,r4,192
c000000000c001dc  49fff051	bl      c000000002bff22c
c000000000c001e0  7c7b1b79	mr.     r27,r3
c000000000c001e4  4082005c	bne     c000000000c00240	# kobject_rename+0xac/0x160
c000000000c001e8  3b800000	li      r28,0
c000000000c001ec  39a00000	li      r13,0
```

Looking at vmlinux:
```
c000000000c001bc:       35 f3 ff 4b     bl      c000000000bff4f0 <kobject_get+0x8>
```
This is 32MB off:
```
>>> 0xc000000002bff4f0 - 0xc000000000bff4f0 
33554432L
```

**Comment by mpe** [comment]
I can't reproduce this locally. How do I build the exact kernel you were testing? And what hardware was that on?

**Comment by shenki** [comment]
I've attached the [config.gz](https://github.com/linuxppc/issues/files/3067046/config.gz). It's one I used a while back, I think it's based on skiroot_defconfig with `s/=m/=y/`.

The build machine is Ubuntu 18.04 with the packaged cross compilers.

**Comment by mpe** [comment]
When `CONFIG_HUGETLB_PAGE=n`, `pmd_huge()` and so on are compiled into static inlines that always return false.

This breaks the logic in `radix__change_memory_range()` and other places, which needs to walk the kernel page tables. The kernel page tables can contain huge pages regardless of whether `CONFIG_HUGETLB_PAGE` is enabled or not.

The result is the page table walking code goes off into the weeds and starts patching random places in memory which it thinks are page tables. One of the things we do is flip `_PAGE_WRITE`, and if you hit the right place in memory that can change the target of a branch, which is what we hit originally.

This patch should catch it in action:
```
diff --git a/arch/powerpc/mm/pgtable-radix.c b/arch/powerpc/mm/pgtable-radix.c
index 931156069a81..5870d2335b90 100644
--- a/arch/powerpc/mm/pgtable-radix.c
+++ b/arch/powerpc/mm/pgtable-radix.c
@@ -12,6 +12,7 @@
 #define pr_fmt(fmt) "radix-mmu: " fmt
 
 #include <linux/kernel.h>
+#include <linux/kallsyms.h>
 #include <linux/sched/mm.h>
 #include <linux/memblock.h>
 #include <linux/of_fdt.h>
@@ -216,6 +217,7 @@ void radix__change_memory_range(unsigned long start, unsigned long end,
 		if (!ptep)
 			continue;
 update_the_pte:
+		BUG_ON(is_kernel_text((unsigned long)ptep));
 		radix__pte_update(&init_mm, idx, ptep, clear, 0, 0);
 	}
 
```

I'll send a fix tomorrow.

**Comment by shenki** [comment]
Nice find! Thanks mpe

**Comment by mpe** [comment]
Temporary fix posted:
https://patchwork.ozlabs.org/patch/1086825/

@kvaneesh is going to look at a proper fix.

**Comment by mpe** [comment]
Fixed in torvalds/linux@d6eacedd1f0ebf00bdf1c77715d194f7c1036fd4, thanks @kvaneesh :1st_place_medal:
